### DISCUSSION TITLE
Constrained iterations, and Datalog-like recursion

### DISCUSSION THREAD

**Discussion by julianhyde** [hydromatic/morel#106] [Ideas]
Morel has successfully integrated many aspects of relational languages into a functional programming language but a good syntax for recursively defined relations has been elusive. This is all the more surprising given that recursive relations are so easily defined in [Datalog](https://en.wikipedia.org/wiki/Datalog), a simpler language based on the deductive paradigm.

We propose a new Morel language concept, named 'constrained iteration' and introduced by a new `suchThat` keyword, that allows relations (including recursive relations) to be defined in a similar way to Datalog. This concept fits elegantly with the rest of Morel, gives Morel the same power as Datalog, and gives the Morel program authors more flexibility in how they structure their programs without sacrificing any power or efficiency.

To solve the mystery of why Datalog is different, we journey through set theory, how mathematics and computer science define functions, and the relationship between fixed-points and recursively-defined functions.

## Two families of data languages 

Deductive data languages (such as Datalog) and data languages based on relational algebra (such as SQL, Morel, [Datafun](http://www.rntz.net/datafun/)) do much the same thing, but go about it in different ways.

Even though their paradigms are different, for many operations, such as joining and filtering, the code in the deductive language (Datalog) will tend have a similar size and structure to the code in the language based on relational algebra (Morel).

When recursion is involved, the languages part ways - Datalog generally seems to be more concise - but first of all, let's explore why programs in the two languages tend to look so similar, most of the time. It goes back to set theory.

In set theory, there are two ways of defining a set: intensional and extensional. Wikipedia gives the [following definitions](https://en.wikipedia.org/wiki/Extensional_and_intensional_definitions):
* An *intensional definition* gives meaning to a term by specifying necessary and sufficient conditions for when the term should be used.
* An *extensional definition* gives meaning to a term by specifying its extension, that is, every object that falls under the definition of the term in question.

Let's suppose you want to make the set of red bicycles, *RB*.

| Type | Definition |
| ---- | ---------- |
| Intensional definition | An object is in *RB* if it is a bicycle and it is red. Furthermore, it is a bicycle if it is a vehicle and has two wheels. |
| Extensional definition | *RB* is the intersection of *R*, the set of red objects, and *B*, the set of bicycles. *B* is the intersection of *V*, the set of vehicles, and *W2*, the set of objects with two wheels. |

These expressions map to programs in Datalog and Morel:

<table>
<tr>
<td>Datalog (intensional)</td>
<td><pre>red_bicycle(x) :- bicycle(x), red(x).<br/><code>bicycle(x) :- vehicle(x), x.wheel_count = 2.</pre></td>
</tr>
<tr/>
<tr>
<td>Morel (extensional)</td>
<td><pre>val bicycles =<br/>  from v in vehicles where v.wheelCount = 2<br/>val redObjects =<br/>  from o in allObjects where o.color = 'red'<br/>val redBicycles = bicycles intersect redObjects</pre></td>
</tr>
</table>

The programs have similar structure, because each operation on an extensional set has a corresponding operator on an intensional set. For example, where Datalog uses "`,`" (boolean AND), Morel uses "`intersect`".

The duality is very clear when we draw the sets in a [Venn diagram](https://en.wikipedia.org/wiki/Venn_diagram):
![Venn diagram](https://user-images.githubusercontent.com/1130553/154903989-ff0fecad-7ec7-476d-8770-a5c71e3bd2db.png)

Venn diagrams are powerful visualizations because, where the dualism is concerned, they don't pick sides. You can interpret the red area intensionally (objects whose color property has the value 'red') and extensionally (the set of all red objects). The duality isn't created by Venn diagrams, or languages such as Datalog or Morel; it has been in set theory from the start.

The relational data model, based on set theory, recognized the duality from the start: [Codd's theorem](https://en.wikipedia.org/wiki/Codd%27s_theorem) proved that domain-independent relational calculus is equivalent to relational algebra. Datalog is a descendant of relational calculus, and has an elegant means to recursively define relations; the descendants of relational algebra have better handling of multisets, aggregation, and nested relations, but their handling of recursive relations is clunky.

## Extents

Observe that the Morel program uses an `allObjects` collection that has no analog in the Datalog program. An extensional language such as Morel needs to start with an *extent* - the set of all objects of a particular type, such as the set of all red objects or the set of all vehicles - whereas an intensional language just needs to define a predicate, saying, in effect, 'hand me an object and I'll tell you whether it is red'.

In practice, extensional programs limit themselves to a finite extent ("the set of vehicles that I own") and mix in some intensional logic ("I know we're working with vehicles here, so give me the set of vehicles and I'll give you back the ones that are red"). But I claim that, by forcing a programmer to choose between intensional or extensional thinking, the language is doing no one any favors. The programmer may have to rewrite code if it turns out that the extent they need is no longer available. And the runtime is forced to execute the program the way the programmer wrote it, even though the other way might be more efficient.

If we could remove the need for this forced switching, would we make it easier to write programs? Let's come back to that question after we look at the one area where deductive and relational languages are significantly different: recursion.

## Recursion

A simple example (based on the [Datafun: a functional Datalog](https://www.cl.cam.ac.uk/~nk480/datafun.pdf) paper) illustrates recursion. The Morel variable `parents` is a relation that contains parent-child relationships:

<pre>
- <b>val parents =
    [("earendil", "elrond"), ("elrond", "elladan"),
      ("elrond", "elrohir"), ("elrond", "arwen")]</b>;
val parents : string * string list
</pre>

How can we write a Morel function that returns the set of (`x`, `y`) pairs such that `x` is an ancestor of `y`? Our first attempt, inspired by SQL's `WITH RECURSIVE` construct, is to write a recursive function that uses `union`:

<pre>
- <b>fun ancestors () =
    (from (x, y) in parents yield (x, x))
    union
    (from (x, y) in parents,
        (y2, z) in ancestors ()
      where y = y2
      yield (x, z))</b>;
val ancestors = fn : unit -> (string * string) list
</pre>

But when we run it we get a stack overflow! It's not surprising when you notice that `ancestors` calls itself recursively with the same arguments that it was invoked with. While it's valid in mathematics to define a function in terms of itself, this is no way to write a function in a programming language.

Our next thought was to teach Morel that the `union` is somehow special, add some notion of monotonicity to the type system, and allow the language to deduce when recursive functions involving `union` proceed towards a fixed point. This seems rather complicated. (How much effort would be required by the user to specify monotonicity? Are there other functions besides `union` that have monotonic behavior? Would we allow users to define their own monotonic functions and types?)

How does Datalog solve the problem? The following program defines an `is_parent` base relation, and a recursively derived `is_ancestor` relation.

<pre>
is_parent(earendil, elrond).
is_parent(elrond, arwen).
is_parent(elrond, elladan).
is_parent(elrond, elrohir).
is_ancestor(X, Y) :- is_parent(X, Y).
is_ancestor(X, Y) :- is_parent(X, Z), is_ancestor(Z, Y).
</pre>

Why is Datalog more concise? Datalog is operating on boolean values (predicates) rather than sets. Whereas the Morel function was invoked with a unit argument and returned a set, Datalog functions were invoked with specific arguments (and invoked many times, once for each value in a set) and returned a boolean.

In mathematics, a relation is a set of n-tuples. A Datalog rule tests membership of a relation. For example, when we invoke `is_ancestor(earendil, arwen)` we are asking whether `(earendil, arwen)` belongs to the relation.

A mathematical function is a special kind of relation - it is a binary relation whose pairs' left elements are unique - and it is well known that in lambda calculus we can define recursive functions using a [fixed-point combinator](https://en.wikipedia.org/wiki/Fixed-point_combinator#Usage_in_programming). It is less often remembered that the 'fixed point' being computed is not the result of a particular call to the function, but the function itself, namely the set of (argument, result) pairs.

If we are going to recursively define relations, then it makes sense to use the same approach. And in fact Datalog does just that. It deduces the contents of a relation by starting with a set of base facts, then applies all possible rules to deduce new facts until it cannot deduce any more. The process converges in a straightforward way because the boolean value - the proof that a tuple is in the relation - is monotonic.

We had trouble making the `ancestors` Morel function recursive because it was trying to work on sets (where monotonicity is not trivial as it is for `bool`) and because each iteration was not just deducing a result for an argument value for which the function was previously undefined.

So, let's abandon that approach. Let's make it possible in Morel to define sets based on a predicate.

## Constrained iteration

We propose a new operator `suchThat`, to make a new clause in the `from` expression. This clause performs what we call 'constrained iteration' (iteration of variables over the values for which a predicate is true), as opposed to the usual iteration over the values of a list- or set-valued expression.

Syntax is as follows:

<pre>
<i>exp</i> &rarr;           (* other expressions *)
  | <b>from</b> [ <i>scan<sub>1</sub></i>  , ... , <i>scan<sub>s</sub></i> ] <i>step</i>*
                              relational expression (s &ge; 0)
<i>scan</i> &rarr; <i>pat</i> [ <b>in</b> | <b>=</b> ] <i>exp</i>
  | <i>id<sub>1</sub></i> , ... , <i>id<sub>i</sub></i> <b>suchThat</b> <i>exp</i>
                              constrained iteration (i &ge; 1)
<i>step</i> &rarr; <b>where</b> <i>exp</i>                filter clause
  | <b>join</b> <i>scan</i> [ <b>on</b> <i>exp</i> ]      join clause
  | <b>group</b> <i>groupKey<sub>1</sub></i> <b>,</b> ... <b>,</b> <i>groupKey<sub>g</sub></i>
    [ <b>compute</b> <i>agg<sub>1</sub></i> <b>,</b> ... <b>,</b> <i>agg<sub>a</sub></i> ]
                              group clause (<i>g</i> &ge; 0, <i>a</i> &ge; 1)
  | <b>compute</b> <i>agg<sub>1</sub></i> <b>,</b> ... <b>,</b> <i>agg<sub>a</sub></i>
                              compute clause (<i>a</i> &gt; 1)
  | <b>order</b> <i>orderItem<sub>1</sub></i> <b>,</b> ... <b>,</b> <i>orderItem<sub>o</sub></i>
                              order clause (<i>o</i> &ge; 1)
  | <b>yield</b> <i>exp</i>
</pre>

The following query computes the set of job titles that are managed by employees in department 20:

<pre>
- <b>fun isManagerOf (e, job) =
    exists (from e2 in emps
      where e2.mgrno = e.empno
      andalso e2.job = job)</b>;
val isManagerOf = fn : Emp * string -> bool
- <b>from e in emps,
      job suchThat isManagerOf (e, job)
    where d.deptno = 20
    yield job</b>;
val it : string list
</pre>

The function `isManagerOf` returns whether employee `e` manages someone with job title `job`.

The clause `job suchThat isManagerOf (e, job)` does something strange. It apparently iterates over all possible jobs, and then checks whether the current employee `e` manages someone with that job.

The set of all possible jobs is infinite, the implementation does not literally iterate over all possible jobs. It drives in the other direction, starting with the set of jobs that actually exist, and constraining `e` to be the manager of people with those jobs. If there is no source of values, there is nothing to drive from, and Morel
rejects the query as unimplementable. For example:

<pre>
- <b>from (e, job) suchThat isManagerOf (e, job)
    yield job</b>;
Error: Argument to suchThat is insufficiently constrained
</pre>

It is OK to define a function that is insufficiently constrained, but you can only call that function from a top-level query if there are sufficient constraints.

When sufficiently constrained, by passing in a value of `e` that is iterating over a finite extent, `isManagerOf` is in effect just defining a 'managerJobs' relation, but is doing it in the 'backwards' Datalog manner of writing a `bool` function.

When defining a relation, there is no preferred direction: sometimes the most natural formulation uses a set-valued expression, in which case you would use forward clause (with `in` or `=`); sometimes it is a boolean-valued expression, in which case you would use a backwards clause (`suchThat`). Morel's optimizer is equally happy with either, and can take a query with a mixture of forwards and backwards clauses and find an efficient plan.

## Recursively defined relations

When defining recursive queries, backwards is the preferred direction. Let's return to the problem of recursively defining the ancestors relation. The definition is much more concise using a 'backwards' function `isAncestor`:

<pre>
- <b>fun isParent (x, y) =
    (x, y) elem parents</b>;
val isParent = fn : string * string -> bool
- <b>fun isAncestor (x, y) =
    isParent (x, y)
    orelse exists (
      from z suchThat isParent (x, z) andalso isAncestor (z, y))</b>;
val isAncestor = fn : string * string -> bool
</pre>

Note how the `z suchThat` clause in the `isAncestor` function introduces a variable `z` and iterates over all possible values of `z` where the boolean expression `isParent (x, z) andalso isAncestor (z, y)` is `true`. We call this 'constrained iteration'.

Constrained iteration is only allowed if Morel can deduce a finite extent for the variable. In this case, the iteration is safe. By looking inside the definition of the `isParent` function, Morel's compiler can deduce that `z` only takes the the values that occur in on the right-hand side of the `parent` relation, namely "elrond", "arwen", "elladan", "elrohir".

One the `isAncestor` function is defined, we can use it in several ways:
<pre>
- <b>isAncestor ("earendil", "arwen")</b>;
val it = true : bool
- <b>isAncestor ("arwen", "elladan")</b>;
val it = false : bool
- <b>from x suchThat isAncestor (x, "arwen")</b>;
val it = ["earendil", "elrond", "arwen"] : string list
- <b>from y suchThat isAncestor ("elrond", y)</b>;
val it = ["elrond", "arwen", "elladan", "elrohir"] : string list
- <b>from (x, y) suchThat  isAncestor (x, y)</b>;
val it =
  [("earendil", "earendil"), ("earendil", "elrond"), ("earendil", "arwen"),
   ("earendil", "elladan"), ("earendil", "elrohir"), ("elrond", "elrond"),
   ("elrond", "arwen"), ("elrond", "elladan"), ("elrond", "elrohir"),
   ("arwen", "arwen"), ("elladan", "elladan"), ("elrohir", "elrohir")]
 : (string * string) list
</pre>

The last of these is the most interesting. By generating both arguments to `isAncestor` in a `suchThat` clause, we have told Morel to generate the entire ancestors relation.

## Related work

[Flix](https://en.wikipedia.org/wiki/Flix_(programming_language)) is another functional language that incorporates Datalog. Datalog-style constraints are first-class values, but they are defined in a sub-language and therefore interoperation with other language features is somewhat limited.

[Datafun](http://www.rntz.net/files/datafun.pdf) is a functional language with relational extensions. It [added recursion](https://www.cl.cam.ac.uk/~nk480/seminaive-datafun.pdf) by making monotonicity a first-class property of every type, and can thereby prove that recursive computations on sets are guaranteed to terminate. Recursive computations are implemented efficiently by rewriting them to use semina&iuml;ve evaluation, the same mechanism used by most Datalog implementations.

## Conclusion

Constrained iteration brings Datalog-like expressive power to Morel, and provides a natural and concise way to recursively define relations.  The feature can be freely mixed with other language features such as function declarations, patterns and polymorphic types, and gives Morel's compiler/optimizer freedom to find the most efficient implementation of a data-parallel program.

Note that constrained iteration and the `suchThat` keyword are proposed language features and are not implemented. Some of the examples require either annotation or inference of record-valued argument types, which is also not implemented.

**Comment by julianhyde** [comment]
I discuss this topic in my talk [Is there a perfect data-parallel programming language?](https://www.slideshare.net/julianhyde/is-there-a-perfect-dataparallel-programming-language-experiments-with-morel-and-apache-calcite). See section 5, 'Functional + Deductive'.

**Comment by Voropay** [comment]
The proposed implementation of recursive queries combines logical and functional semantics, which, in my opinion, is not easy to implement and use. 
The same function definition must act as standard function that deterministically describes the sequence of computations in one cases, and as a logical query that can non-deterministically derive all possible values of its variables in other cases. This means that the function must be able to enumerate all possible values ​​of its arguments, support inverse calculations, and so on. 
In your examples, in the case of isAncestor, the compiler can understand that the variable values ​​are associated with the set of isParent values and enumerate them. But for isManagerOf, it can't, even though we could logically deduce that job is limited to the emps.job. 
In addition, it will be necessary to decide on narrowing (variables binding) and non-deterministic computation strategies. There are a few hybrid functional logic programming languages, such as Curry and Mercury, that try to handle this, but in my opinion they are all too complex for practical use.

An alternative approach might be to consider separating concepts of queries and functions. 
So, the functions will be ordinary functions describing a deterministic order of calculations. And the queries will be able to perform logical search and return data rows. Query definitions will have a clear structure, including a name, a list of their attributes, data sources, join conditions, filter predicates, etc. Then the compiler will be able to easily extract all the necessary information for their efficient execution. 
Using queries we could model relationships between tables or objects collections in relational manner and apply functions to process the query results. But this will be a different language with different semantics and for different purposes.

**Reply by julianhyde** [reply]
@Voropay, Thank you for the feedback! I really appreciate hearing from someone who is trying to solve similar challenges.

Can you elaborate what features of Curry and Mercury made them difficult to use? I would like to avoid those pitfalls if possible.

If you concern was that functions need to declare their available modes (as in Mercury), my plan was to only require a boolean function to be invertible if it is applied it to one or more unconstrained variables. Thus the use of the function is an ask for inversion, as opposed to the function advertising that it is invertible. 

Before I started reading about mode-inversion in Prolog, I was not familiar with idea of deterministic order, e.g. Mercury's declaration
```
:- mode append(in, in, out) is det.
:- mode append(out, out, in) is multi
```
Database users are familiar with nondeterminism, even though they don't call it that. They know that iterating through the employees in a department may occur in any order, and that each employee has at most one manager. These notions are captured by the type system (e.g. the departments field in the employees table has type `bag`, indicating nondeterminisic order, and the `manager` field has type `employee option`)

> But for isManagerOf, it can't, even though we could logically deduce that job is limited
> to the emps.job.

The reason we can't invert `isManagerOf` is because we don't know the extent of the `employee` type. There might be more than one table with employees. This makes the function more useful -- because it can apply to any employees table -- but it requires the caller to ground the query in a particular employees table.

> An alternative approach might be to consider separating concepts of queries and functions.

I see many benefits keeping queries and functions together. Morel has a logical data model, and therefore does not distinguish between persistent tables and collections in memory (both may be ordered or unordered, as indicated by the types `list` and `bag`). What a database user would call a "query" is just an expression that involves one or more collections. Any expression, including a query, can be parameterized to create a a function, and this allows you to decompose your application into smaller units.

The data model includes types (built-in, anonymous record types, and user-defined types including sum types) and types may have constraints (boolean expressions that must be true for an instance of the type to be valid). Via constraints we can create unique keys, foreign keys and therefore model relationships.

Databases and logic systems have for a long time treated the boolean type differently from all other types. Users of general-purpose languages find that distinction unnatural and limiting. For database users, there are often two ways to write a query -- via predicates or algebra -- but one might be more concise than the other. (I was particularly thinking of recursive queries, aka fixed point queries, where predicates are more concise than `iterate` ... `union`.) With inversion, I want to let people people write it whichever way makes most sense, and use it whichever way makes most sense.

**Comment by Voropay** [comment]
@julianhyde, Thanks for your reply! 
I'll try to express my thoughts in more detail:
If I understand correctly, your logic for integrating queries and functions looks like this:
`outer query => suchThat construct => Boolean function => exists predicate => inner query`
You've decided to make functions the main, reusable element of your program, and at the same time, you also need functions to have query semantics. Therefore, in order to integrate functions into queries you introduced the suchThat construct, which creates a query based on a function and binds the function's arguments to the query context.
This way, you separate the data extraction logic (the logical predicate describing the relationships between the source data) from the rest of the query and format it as a function. At the same time, this function can be used as a regular function. 

This certainly makes sense, but I have the following comments:
1) While queries are described using relational algebra (from, join, where, yield), functions are constructed in the style of logical relational calculus (as predicates). Although their mathematical foundations are equivalent, their simultaneous use creates stylistic inconsistency and confusion. I think it would be nice, if possible, to unify the style. I would prefer to sacrifice the interchangeability of functions and logical predicates in order to unify the style, but that's my personal preference.
2) Unfortunately, not all functions can be converted into queries, only boolean and inversible ones. Furthermore, some arguments must be grounded in a specific data source, while for others, the compiler can do this automatically. This also creates confusion about which functions can be used and which cannot, and which arguments must be grounded and which must not. As a user, I would prefer such functions to be separated from the rest, so that I can clearly see which functions I can convert to queries and which arguments need to be grounded and which do not. This should be explicitly specified in function definition either using special types, or keywords, modifiers, generic parameters, etc.
3) I suppose that implementing function inversibility is a tricky task, and inverting only Boolean functions is not enough. For example, the function 
`recordsByDate(record: SomeRow, dateStr: String) = SomeDataTable(record) and record.date = parseDateStr(dateStr) `
is inversible only if the 
`parseString(dateStr: String): Date `
is inversible too. Also, you will need a fallback mechanizm to perform a full search if the predicate function cannot be inversed.
4) The language should be fairly simple because it's users will be mostly data engineers who are not familiar with the complex concepts of logical and functional programming. My main concern is that Curry and Mercury are poorly suited for mainstream use because their barrier to entry is too high and they are too flexible for data processing tasks. You'll have to hide all their complexity under the hood, providing users with only the abstractions they need.

I also tried (and am still trying) to create a convenient and modern language for working with data, but in a different way - not with functional programming but rather in the style of object-oriented logic. 
The functional style manages computational complexity well. It can decompose a complex algorithm into smaller parts and then combine them in many flexible ways. This makes sense in cases where computation logic is the main focus such as scientific data processing, data integration (for example, fraud detection), maybe custom machine learning models, etc.
But when working with data, data structures are often important as well, especially in analytics where the main focus is the clarity of data assets and their dependencies. In ELT pipelines, each step is a data object, the description of which includes its structure (a list of fields) and the logic for obtaining the data (a query to other data objects or an algorithm for generating them). I am trying not to break these descriptions into schemas and functions, but to find a way to work with them as a single whole. To achieve this, I borrowed OOP techniques for code reuse, such as inheritance, parameterization, and introduced special types of composable queries to describe metrics, attributes, relationships, and so on. My goal is to bring the program code closer to the logical data model.

Although our approaches have different ideologies, they overlap in many ways and can complement each other. So I'll definitely be keeping an eye on your work on Morel!
